### PR TITLE
N-Queen Problem

### PR DESCRIPTION
**Pull request by Djsmk123**
# Welcome to Dart community

### ***Updates***

*[ ] Added N Queen Problem solution 
*[ ] All test cases Passed

### **Checklist:**
* [ ] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Dart/blob/master/CONTRIBUTING.md).
* [ ] This pull request is all my own work -- I have not plagiarized.
* [ ] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Dart files are placed inside an existing directory.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: backtracking/n-queen.dart [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+on an n x n chessboard that no queens attack each other by being in the same row, column or diagonal
+We have to find maximum no of solutions that are possible.
+
+.*/
+
+// board
+List<List<int>> board = [];
+
+// checking is it safe place for a queen.
+bool isSafe(int n, int x, int y) {
```
Comment (akashgk) [inline backtracking/n-queen.dart:16]: can you use descriptive variable names instead of x, y, n etc

File: backtracking/n-queen.dart [hunk 2]
```diff
@@ -0,0 +1,119 @@
+import 'dart:io';
+import 'package:test/test.dart';
+
+/*N-Queens Problem
+
+N - Queens problem is to place n - queens in such a manner 
+on an n x n chessboard that no queens attack each other by being in the same row, column or diagonal
+We have to find maximum no of solutions that are possible.
+
```
Comment (akashgk) [inline backtracking/n-queen.dart:9]: Can you add link to some wikipedia or GFG articles?

File: backtracking/n-queen.dart [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+void printBoard(int n) {
+  for (int i = 0; i < n; i++) {
+    for (int j = 0; j < n; j++) {
+      stdout.write(board[i][j]);
+    }
+    stdout.write("\n");
+  }
+  stdout.write("\n");
```
Comment (akashgk) [inline backtracking/n-queen.dart:54]: I think we can avoid print statements. just the test cases would be fine.

File: backtracking/n-queen.dart [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+int SolveNQueen(int i, int n) {
```
Comment (akashgk) [inline backtracking/n-queen.dart:57]: small letter camelCase function name

File: backtracking/n-queen.dart [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+      ways += SolveNQueen(i + 1, n);
+
+      //backtrack
+      board[i][j] = 0;
+    }
+  }
+  return ways;
+}
+
+int main() {
```
Comment (akashgk) [inline backtracking/n-queen.dart:77]: no real need for ```int main()``` and returning 0;

**Review by akashgk — CHANGES_REQUESTED** [review]

File: backtracking/n-queen.dart (hunk 1 above)
Comment (Djsmk123) [reply backtracking/n-queen.dart:16]: Sure I will  change

File: backtracking/n-queen.dart [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    equals(10);
+  });
+  test("N Queen #testcase 6", () {
+    board = List.generate(20, (i) => List.generate(20, (j) => 0));
+    solveNQueen(0, 6);
+    equals(4);
+  });
+  test("N Queen #testcase 7", () {
+    board = List.generate(20, (i) => List.generate(20, (j) => 0));
+    solveNQueen(0, 7);
```
Comment (akashgk) [inline backtracking/n-queen.dart:104]: Some of the test are failing. Can you please check the correctness of the algorithm?

**Review by akashgk — CHANGES_REQUESTED** [review]

File: backtracking/n-queen.dart [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+      //backtrack
+      board[row][col] = 0;
+    }
+  }
+  return ways;
+}
+
+void main() {
+  test("N Queen #testcase 1", () {
+    board = List.generate(20, (i) => List.generate(20, (j) => 0));
```
[the remark this answers is not in the document]
Comment (Djsmk123) [reply backtracking/n-queen.dart]: Added comment in the code why we are generating board list.

**Review by akashgk — CHANGES_REQUESTED** [review]
LGTM, Please make the requested changes.

**Review by akashgk — CHANGES_REQUESTED** [review]
@Djsmk123 The Algorithm does not seem to be correct. Can you reverify the logic. 

Problem statement is to place N Queens on N * N board. But the board is always a 20 * 20 board. and we are varying the size.

**Comment by Djsmk123** [comment]
We know that This problem will be take too much time in case of N>20 so It is like maximum value. 
I can make it N*N.

**Review by akashgk — APPROVED** [review]

**Pull request merged**